.github/CONTRIBUTING.md

```diff
@@ -341,3 +450,124 @@ def clear_course_nicknames(self):
+checks on code to identify and fix potential `TypeError` bugs before runtime.
+Type annotations should be added to maintain consistency across modules.
+
+#### Type hint formatting
+
+Type annotations follow the general format below to define expected types for
+method parameters and returns.
+
+```python
+def my_function(param1: str, param2: str) -> str:
+    return f'You submitted {param1} and {param2}!"
+```
+
+#### Type Hints in Classes
+
+Classes instantiated with an `__init__()` method are expected to return `None`.
+
+```python
+class MyClass():
+
+    __init__(self, param1: str, param2: int) -> None:
+        self.param1: param1,
+        self.param2: param2
+```
+
+Class methods, on the other hand, can be annotated with a specific return type.
+In the example below, the `get_course()` method returns an instance of `Course`:
+
+```python
+class Canvas():
+
+    def get_course(
+        self,
+        course: Union[int, Course],
+        use_sis_id: bool = False,
+        **kwargs: Optional[dict],
+    ) -> Course: ...
+```
+
+Often, a method can have multiple valid parameters. In `get_course`, the
+`course` param can be either an `int` or a `Course` object. The special `Union`
+type allows you to accept multiple types for a given parameter. Every method has
+an optional `**kwargs` argument which can be annotated by importing the
+`Optional` type from the `typing` package:
+
+```python
+from typing import Optional, Union
+```
+
+Both types accept a list of _other_ types which would be valid in the function:
+
+```python
+def my_function(param1: Union[str, int], **kwargs: Optional[str, int]) -> Any: ...
+```
+
+#### Python 3.6 type hints
+
+Modules are often imported when necessary in specific class methods rather than
+at the top of the file. This can cause problems because the imports happen as
+the code is executed, not when the program is initialized. This will cause
+errors to show when the function is returning a valid value:
+
+```python
+"""
+In this example, <AppointmentGroup> would have an error in the static check because
+it has not been imported before the function definition. However, this is a valid
+function because it is instantiated by the time the function returns.
+"""
+def get_appointment_group(
+        self, appointment_group: Union[int, AppointmentGroup], **kwargs: Optional[dict]
+    ) -> AppointmentGroup:
+
+        from canvasapi.appointment_group import AppointmentGroup
+
+        appointment_group_id = obj_or_id(
+            appointment_group, "appointment_group", (AppointmentGroup,)
+        )
+
+        response = self.__requester.request(
+            "GET",
+            "appointment_groups/{}".format(appointment_group_id),
+            _kwargs=combine_kwargs(**kwargs),
+        )
+        return AppointmentGroup(self.__requester, response.json())
+```
+
+To get around this problem, the `typing` package has a `TYPE_CHECKING` module
```

Comment:
Is it possible to use [forward references](https://www.python.org/dev/peps/pep-0484/#forward-references) instead of conditional importing? I notice you [use them later](https://github.com/ucfopen/canvasapi/pull/435/files#diff-98e64bc1cd2db9333c6effe87bbe0d6dfe8714aba4c6bde45aa037fe0796e44cR560), but I'm not sure if the conditional imports are a requirement for forward references to work or not.

Reply:
I'd tried that, but if I remember correctly, I had inconsistent results...mypy still threw errors. But, I could be remembering that wrong. I'll pull the last update back into this branch and see what happens.